### ISSUE TITLE
debugging information removed from static library with -g option (or -C debuginfo=2)

### ISSUE THREAD

**Issue by mikebenfield** [rust-lang/rust#45218]
If I assemble an assembly file and put the object into a static library, then link with a  Rust program using the `-g` option, `rustc` removes some debugging information about my assembly file. Here's an example *without* the `-g` option, where things work as they should. (This is on macOS. I think this will run on Linux too, but I haven't tested there. This program won't run correctly on Windows due to the ABI.)

```
$ cat main.rs
extern "C" {
    fn asm_function();
}


fn main() {
    println!("Hello, world!");
    unsafe {
        asm_function();
    }
}
$ cat file2.s
.intel_syntax noprefix

.text
        .global _asm_function, asm_function

asm_function:
_asm_function:
        lea     rdi, [rip + message]
        sub     rsp, 8
        call    _puts
        add     rsp, 8
        ret

.data
message:
    .asciz "asm function"
$ clang -g -c file2.s -o file2.o
$ ar crus libfile2.a file2.o
$ rustc -L . -l file2 main.rs -o out
$ lldb out
(lldb) target create "out"
Current executable set to 'out' (x86_64).
(lldb) breakpoint set --file file2.s --line 9
Breakpoint 1: where = out`asm_function + 7, address = 0x00000001000015d7
```

OK, that's fine. But if I give `rustc` the `-g` option:
```
$ rustc -g -L . -l file2 main.rs -o out
$ lldb out
(lldb) target create "out"
Current executable set to 'out' (x86_64).
(lldb) breakpoint set --file file2.s --line 9
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
```

LLDB no longer knows about `file2.s`. Oddly, if I write things in C instead of assembly, it seems this problem doesn't occur.

**Comment by jethrogb** [comment]
Works fine for me on Linux when using `gcc` and `gdb`. Note you have to compile with `-fPIC` and change the symbol on `file2.s:10` to `puts@PLT`.

**Comment by mikebenfield** [comment]
Right. I carefully used both labels `asm_function` and `_asm_function`, hoping that it would work on Linux, and completely forgot about `_puts`. Brilliant.

**Comment by wesleywiser** [comment]
Visited during wg-debugging triage. Based on the discussion history, it sounds like this was resolved and (testing on Linux with Rust 1.61 no longer repros).
